### PR TITLE
[CodeComplete] Use PrintOptionalAsImplicitlyUnwrapped to print IUO

### PR DESCRIPTION
**Pull request by rintaro**
Other instances of fb9c65e. Consistently use `PrintOption.PrintOptionalAsImplicitlyUnwrapped` to print IUO.

rdar://problem/41046225
rdar://problem/42443512

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/IDE/CodeCompletionResultBuilder.h [hunk 1]
```diff
@@ -372,13 +372,8 @@ class CodeCompletionResultBuilder {
 
     PrintOptions PO;
     PO.SkipAttributes = true;
-    std::string TypeName;
-    if (IsIUO) {
-      assert(Ty->getOptionalObjectType());
-      TypeName = Ty->getOptionalObjectType()->getStringAsComponent(PO) + "!";
```
Comment (rintaro) [inline lib/IDE/CodeCompletionResultBuilder.h]: In theory, this shouldn't crash. But this crash was caused by
https://github.com/apple/swift/blob/e9cb62d476496787ccc92c506e9059db58c96614/lib/IDE/CodeCompletion.cpp#L2177-L2180

`expectedNumParams != BodyParams->size()` is not enough.

In fact, this fix still doesn't produce correct result, but at least, it fixes the crashes. I want to include this fix for 4.2. I will fix this `FIXME` for later releases.

File: lib/IDE/CodeCompletion.cpp [hunk 2]
```diff
@@ -2481,16 +2481,17 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
         if (ResultType->isVoid()) {
           OS << "Void";
-        } else if (!IsImplicitlyCurriedInstanceMethod
-                   && FD->getAttrs().hasAttribute<ImplicitlyUnwrappedOptionalAttr>()) {
+        } else {
           // As we did with parameters in addParamPatternFromFunction,
           // for regular methods we'll print '!' after implicitly
           // unwrapped optional results.
-          auto ObjectType = ResultType->getOptionalObjectType();
-          OS << ObjectType->getStringAsComponent();
```
Comment (rintaro) [inline lib/IDE/CodeCompletion.cpp]: I couldn't find a reproducer for this crash. But we have several logs crashed in here. It means, result type is *not* optional even when the function decl has `ImplicitlyUnwrappedOptionalAttr`.
@rudkx Are you able to come up with a possibility that causes this situation?

Comment (rudkx) [reply lib/IDE/CodeCompletion.cpp]: The only thing I can think of is that this is related to the adjustment of `FunctionType` earlier in the function.

File: lib/IDE/CodeCompletion.cpp [hunk 3]
```diff
@@ -2481,16 +2481,17 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
-          OS << "!";
-        } else {
-          ResultType.print(OS);
+          bool IsIUO =
+              !IsImplicitlyCurriedInstanceMethod &&
+              FD->getAttrs().hasAttribute<ImplicitlyUnwrappedOptionalAttr>();
+
+          PrintOptions PO;
+          PO.PrintOptionalAsImplicitlyUnwrapped = IsIUO;
+          ResultType.print(OS, PO);
```
Comment (rudkx) [inline lib/IDE/CodeCompletion.cpp]: This fix is going to cause other problems. All optionals will be printed with `!`, so if e.g. if you try your change with a function that returns `[Int?]!`, I think you'll find it prints as `[Int!]!`.

This applies to the earlier fix for parameters as well.

Comment (rintaro) [reply lib/IDE/CodeCompletion.cpp]: I don't think so. This option only applies to the first Optional found.
https://github.com/apple/swift/blob/master/lib/AST/ASTPrinter.cpp#L3748-L3754

I will add a test case for it.

Comment (rudkx) [reply lib/IDE/CodeCompletion.cpp]: Ah yes, you're right, I had forgotten the code ensured that.

**Review by rudkx — APPROVED** [review]

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
Resolved conflict.
@swift-ci Please smoke test

**Pull request merged**